File: android/src/main/kotlin/com/microsoft/clarity/api/ClarityNetworkHostApiImpl.kt
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

package com.microsoft.clarity.api

import android.content.Context
import android.os.Handler
import android.os.Looper
import com.microsoft.clarity.bridges.NetworkBridge
import com.microsoft.clarity.generated.ClarityNetworkHostApi
import com.microsoft.clarity.generated.ConnectivityChangedEventData
import com.microsoft.clarity.generated.ConnectivityChangedStreamHandler
import com.microsoft.clarity.generated.ConnectivityType
import com.microsoft.clarity.generated.PigeonEventSink
import io.flutter.plugin.common.BinaryMessenger

internal class ClarityNetworkHostApiImpl(
    private val networkBridge: NetworkBridge,
) : ClarityNetworkHostApi {
    override fun getConnectivityStatus(): List<ConnectivityType> = networkBridge.getNetworkTypes()
}

internal class ConnectivityChangedStreamHandlerImpl(
    private val networkBridge: NetworkBridge,
) : ConnectivityChangedStreamHandler() {
    private val mainHandler = Handler(Looper.getMainLooper())

    override fun onListen(
        p0: Any?,
        sink: PigeonEventSink<ConnectivityChangedEventData>,
    ) {
        networkBridge.startListening { types ->
            mainHandler.post {
                sink.success(ConnectivityChangedEventData(types = types))
            }
        }
    }

    override fun onCancel(p0: Any?) {
        networkBridge.stopListening()
    }

    fun dispose() {
        networkBridge.stopListening()
    }

    companion object {
        fun register(
            messenger: BinaryMessenger,
            networkBridge: NetworkBridge,
        ): ConnectivityChangedStreamHandlerImpl {
            val handler = ConnectivityChangedStreamHandlerImpl(networkBridge)
            ConnectivityChangedStreamHandler.register(messenger, handler)
            return handler
        }
    }
}

File: ios/clarity_flutter/Sources/clarity_flutter/api/ClarityNetworkHostApiImpl.swift
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import Foundation

internal class ClarityNetworkHostApiImpl: ClarityNetworkHostApi {

    private let networkBridge: NetworkBridge

    init(networkBridge: NetworkBridge) {
        self.networkBridge = networkBridge
    }

    func getConnectivityStatus() throws -> [ConnectivityType] {
        return networkBridge.getNetworkTypes()
    }
}

internal class ConnectivityChangedStreamHandlerImpl: ConnectivityChangedStreamHandler {

    private let networkBridge: NetworkBridge

    init(networkBridge: NetworkBridge) {
        self.networkBridge = networkBridge
    }

    override func onListen(
        withArguments arguments: Any?,
        sink: PigeonEventSink<ConnectivityChangedEventData>
    ) {
        networkBridge.startListening { [weak sink] types in
            DispatchQueue.main.async {
                sink?.success(ConnectivityChangedEventData(types: types))
            }
        }
    }

    override func onCancel(withArguments arguments: Any?) {
        networkBridge.stopListening()
    }

    func dispose() {
        networkBridge.stopListening()
    }
}
